File: frontend/types/institution-skill-gap.ts
// Mirrors backend/app/schemas/institution_skill_gap.py.
//
// An APPLICATION-scoped skill gap tool, not an institution-wide dashboard:
// student -> application -> job/internship -> required skills -> student
// skills -> gap. The match fields (score/recommendation/skill_coverage/
// matched_skills/needs_improvement_skills/missing_skills) reuse the EXACT
// same `MatchSkill` / `MatchRecommendation` shape Industry's own
// ApplicationMatchResponse already defines (types/application.ts) --
// imported, not redefined, so the institution and industry views can never
// silently drift into two different skill-match shapes.

import type {
  ApplicationStatus,
  MatchRecommendation,
  MatchSkill,
  OpportunityType,
} from "@/types/application";

export interface SkillGapApplicationSummary {
  application_id: string;
  student_id: string;
  full_name: string | null;
  username: string | null;
  opportunity_type: OpportunityType;
  opportunity_title: string | null;
  company_name: string | null;
  status: ApplicationStatus;
  applied_at: string | null;
  score: number;
  recommendation: MatchRecommendation;
  skill_coverage: string;
  matched_count: number;
  needs_improvement_count: number;
  missing_count: number;
}

export interface SkillGapListResponse {
  applications: SkillGapApplicationSummary[];
}

/** The student's own recorded skill -- not just the subset a particular
 * opportunity requires. Same shape as the Student Directory's
 * StudentSkillSummary (backend/app/schemas/institution_student.py). */
export interface StudentSkillEntry {
  skill_name: string;
  proficiency_level: string;
  is_verified: boolean;
}

export interface SkillGapDetail extends SkillGapApplicationSummary {
  required_count: number;
  matched_skills: MatchSkill[];
  needs_improvement_skills: MatchSkill[];
  missing_skills: MatchSkill[];
  student_skills: StudentSkillEntry[];
}

export interface SkillGapListParams {
  search?: string;
  opportunity_type?: OpportunityType;
  status?: ApplicationStatus;
}

File: backend/app/schemas/institution_skill_gap.py
"""Pydantic schemas for the Institution Skill Gap Analysis
(backend/app/api/institution.py, GET /institution/skill-gaps,
GET /institution/skill-gaps/{application_id}).

This is an APPLICATION-SPECIFIC skill gap tool, not an institution-wide
prediction dashboard: for one of the institution's own linked students'
applications, how does that student's recorded `student_skills` compare to
the opportunity's own `internship_skills` / `job_skills` requirements.

The match fields (`score`, `recommendation`, `skill_coverage`,
`matched_skills`, `needs_improvement_skills`, `missing_skills`) are the
EXACT same shape Industry's own ApplicationMatchResponse already uses
(app.schemas.application) -- `MatchSkill` and `MatchRecommendation` are
imported, not redefined, so a client (or a test) can never observe two
different skill-match shapes across roles. `student_skills` reuses
`StudentSkillSummary` from app.schemas.institution_student for the exact
same reason -- the Institution Student Directory already defined "how a
student's skill list is shaped for institution eyes"; this module does not
invent a second one.
"""

from pydantic import BaseModel

from app.schemas.application import MatchRecommendation, MatchSkill, OpportunityType
from app.schemas.institution_student import StudentSkillSummary


class SkillGapApplicationSummary(BaseModel):
    """One row in the Skill Gap Analysis list -- one application by one of
    the institution's own linked students, with its deterministic skill
    match already computed (app.services.match_service.compute_match)."""

    application_id: str
    student_id: str
    full_name: str | None
    username: str | None
    opportunity_type: OpportunityType
    opportunity_title: str | None
    company_name: str | None
    status: str
    applied_at: str | None
    score: int
    recommendation: MatchRecommendation
    skill_coverage: str
    matched_count: int
    needs_improvement_count: int
    missing_count: int


class SkillGapListResponse(BaseModel):
    applications: list[SkillGapApplicationSummary]


class SkillGapDetailResponse(SkillGapApplicationSummary):
    """Everything the list row has, plus the full skill breakdown and the
    student's own complete recorded skill list (not just the skills the
    opportunity happens to require)."""

    required_count: int
    matched_skills: list[MatchSkill]
    needs_improvement_skills: list[MatchSkill]
    missing_skills: list[MatchSkill]
    student_skills: list[StudentSkillSummary]
